pkg/blocks/step.go

```diff
@@ -22,241 +22,176 @@ package blocks
-func (a *Act) Validate() error {
-	// Make sure name is of format we can index
-	if strings.Contains(a.Name, " ") {
-		return errors.New("name must not contain spaces")
+	// Decode all of the shared fields.
+	// Use of this auxiliary type prevents infinite recursion
+	var csf CommonStepFields
+	err := node.Decode(&csf)
+	if err != nil {
+		return err
 	}
+	s.CommonStepFields = csf
 
-	return nil
-}
+	if s.Name == "" {
+		return errors.New("no name specified for step")
+	}
+
+	// figure out what kind of action is
+	// associated with executing this step
+	s.action, err = s.ParseAction(node)
+	if err != nil {
+		return err
+	}
 
-// CheckCondition checks the condition specified for an Act and returns true
-// if it matches the current OS, false otherwise. If the condition is "always",
-// the function returns true.
-// If an error occurs while checking the condition, it is returned.
-//
-// **Returns:**
-//
-// bool: true if the condition matches the current OS or the
-// condition is "always", false otherwise.
-//
-// error: An error if an error occurs while checking the condition.
-func (a *Act) CheckCondition() (bool, error) {
-	switch a.Condition {
-	case "windows":
-		if runtime.GOOS == "windows" {
-			return true, nil
+	// figure out what kind of action is
+	// associated with cleaning up this step
+	if csf.CleanupSpec.IsZero() {
+		// hack for subTTPs - they should always use their default cleanup
+		if ShouldUseImplicitDefaultCleanup(s.action) {
+			s.cleanup = s.action.GetDefaultCleanupAction()
 		}
-	case "darwin":
-		if runtime.GOOS == "darwin" {
-			return true, nil
+	} else {
+		useDefaultCleanup, err := isDefaultCleanup(&csf.CleanupSpec)
+		if err != nil {
+			return err
 		}
-	case "linux":
-		if runtime.GOOS == "linux" {
-			return true, nil
+		if useDefaultCleanup {
+			if dca := s.action.GetDefaultCleanupAction(); dca != nil {
+				s.cleanup = dca
+				return nil
+			} else {
+				return fmt.Errorf("`cleanup: default` was specified but step %v is not an action type that has a default cleanup action", s.Name)
+			}
 		}
-	// Run even if a previous step has failed.
-	case "always":
-		return true, nil
 
-	default:
-		return false, nil
+		s.cleanup, err = s.ParseAction(&csf.CleanupSpec)
+		if err != nil {
+			return err
+		}
 	}
-	return false, nil
+	return nil
 }
 
-// MakeCleanupStep creates a CleanupAct based on the given yaml.Node.
-// If the node is empty or invalid, it returns nil. If the node contains a
-// BasicStep or FileStep, the corresponding CleanupAct is created and returned.
-//
-// **Parameters:**
-//
-// node: A pointer to a yaml.Node containing the parameters to
-// create the CleanupAct.
-//
-// **Returns:**
-//
-// CleanupAct: The created CleanupAct, or nil if the node is empty or invalid.
-//
-// error: An error if the node contains invalid parameters.
-func (a *Act) MakeCleanupStep(node *yaml.Node) (CleanupAct, error) {
-	if node.IsZero() {
-		return nil, nil
-	}
-
-	basic, berr := a.tryDecodeBasicStep(node)
```

Comment:
we only supported two types of cleanup actions - this refactor now allows us to support all of them, as intended and proven by the newly-added tests